proposed/0000-crust-refactor/0000-crust-refactor.md

```diff
@@ -0,0 +1,501 @@
+cancel the connection attempt after some time period, or if another event
+happens first, or if the application is shutting down, or any number of other
+reasons. To enable this we can instead make the connection happen in two
+stages. First we create a pair of objects, one of which is used to perform the
+blocking call, the other of which is used to cancel the blocking call.
+
+```rust
+let (connect_go, connect_kill) = Stream::connect(addr);
+```
+
+We forward `connect_kill` to the part of the program that will be in charge of
+cancelling the call and then perform the call.
+
+```rust
+sender.send(connect_kill)
+let stream = connect_go.go();
+```
+
+Alternatively, `connect_go` could be sent to another thread to perform the
+connection, the important part is that the `go` and `kill` are in two seperate
+places of control. To cancel the operation we simply drop `connect_kill`.
+
+As this is a common pattern throughout the libraries, there's a trait for it.
+
+```rust
+trait Go {
+    type Output;
+
+    fn go(self) -> Option<Self::Output> { .. }
+    fn go_timeout(self, timeout: Duration) -> Option<Self::Output> { .. }
+    fn go_deadline(self, deadline: SteadyTime) -> Option<Self::Output> { .. }
+    fn go_opt_timeout(self, opt_timeout: Option<Duration>) -> Option<Self::Output> { .. }
+    fn go_opt_deadline(self, opt_deadline: Option<SteadyTime>) -> Option<Self::Output>
+}
+```
+
+Wherever types named `*Go` and `*Kill` occur in the libraries, this is the
+pattern that's being used. Additionally, it's sometimes necessary to control a
+blocking operation while it is taking place. In these cases, a `*Go` and
+`*Controller` type are used where the controller is used to control the object
+performing the operation. As with `*Kill`, dropping `*Controller` unblocks the
+operation. For an example of this see the `Beacon::recv` method.
+
+For operations that can be performed repeatedly the `*Go` type will return
+itself upon success. For example, the `Output` of `ReadGo` is `(ReadGo, usize)`
+
+To see how functions that use this blocking style can be used and composed
+consider an example where a user wants to write a method which
+
+ (0) Connects to an endpoint.
+ (1) Writes "ping" to the stream.
+ (2) Reads data from the stream.
+ (3) Prints the result.
+
+First, let's consider a simple example where we want to do these things in
+sequence and timeout the whole operation after 1 second;
+
+```rust
+pub fn ping(endpoint: Endpoint) {
+    let deadline = SteadyTime::now() + Duration::from_secs(1);
+
+    let (connect_go, _k) = Stream::connect(endpoint);
+    let stream = connect_go.go_deadline(deadline);
+
+    let (write_go, _k) = stream.write(b"ping");
+    let _ = write_go.go_deadline(deadline);
+
+    let mut data = [0u8; 256];
+    let (read_go, _k) = stream.read(&mut data[..]);
+    let n = read_go.go_deadline(deadline);
+
+    println!("data == {:?}", data);
+}
+```
+
+Simple, right? For a richer example, consider how we implement this procedure
```

Comment:
I see this as where the rotor design seems to fit nicely with enum variants arranged in a state machine pattern. There it seems the time-out and any other context may be passed to child state machines. I am not saying we should do this, but it is interesting to contrast these patterns and as you have said it would be interesting to do a minimal example of each. Interesting never the less.